### ISSUE TITLE
How about unpacking the app?

### ISSUE THREAD

**Issue by dscham** [VictorDrijkoningen/trifo-robotics-rev-eng#4]
Couldn't we get the Trifo Home app and unpack it. To find out how the 0MQ connection works, at least. Maybe there even is the SSH key.

Heads up though: I have no idea of APK unpacking/decompiling. But it seems like the next logical step, before possibly bricking the device in disassembly.

**Comment by FrankRizzo890** [comment]
I've had good success with THIS tool:  https://github.com/charles2gan/GDA-android-reversing-Tool

And yes, ripping apart the Android app IS a good next step.

**Comment by VictorDrijkoningen** [comment]
I have already looked into this a little bit, and my inexperienced findings are that the app does not locally connect to the vacuum.  I think this is also confirmed by the app being able to connect to the vacuum when not in the same network.

What i did was; use adb to get the apk off my phone. Then i unpacked it with some tool i forgot the name of. Then i looked into all the files with keywords such as: 'ssh', 'password', 'key' etc. Under my knowledge, this did not reveal any way to access the vacuum. I did uncover some hard coded passwords, but one set was for MQTT, which looked like it was used for testing, but I'm not sure. The other hard coded password was for accessing a dev menu in the app. I accessed this menu and i could not see anything that would be of value to my efforts. 

I have not uploaded the ripped app from my phone, as I'm not sure about the legality / personal data it holds.

**Comment by VictorDrijkoningen** [comment]
>before possibly bricking the device in disassembly.

Oh yeah, and the device i have (MAX V1.1) is very much easily opened and disassembled, as i have done this a few times to see if there are usable uart ports etc. 
TLDR for the uart ports, with my limited experience, i could only read on one of the uart ports, but not write.

**Comment by dscham** [comment]
Well, I kinda induced desoldering the flash and or reading it in circuit in my disassembly shorthand 😅 

I actually got a Waveshare UART/SPI/JTAG to USB, being inspired by this. But I didn't get around to using it yet. Don't think it will be of much help anyways. According to your struggles. But still a good tool to have.

**Comment by FrankRizzo890** [comment]
> I did uncover some hard coded passwords, but one set was for MQTT

OK, so the app probably drops events into an MQTT server (it "publishes" events I believe it the proper term), and the vac connects to that server and receives those messages (it "subscribes" to those events).  

There *COULD* be all the info you need right there.  *IN THEORY* you have the server's address, the login credentials, and the events that are sent.  Using a local proxy, and redirecting their server to YOUR server in your router, you might be able to gain "local control" using these.

I used THIS (https://github.com/eclipse/mosquitto) for a local server when I was doing some MQTT development.  It's small, easy to config, and has apps to let you read/write (Sub/Pub) messages to the server.

**Comment by dscham** [comment]
First time look at decompiled stuff. It's really overwhelming. But the AllUris and AppStrings pages in GDA look like a good starting point. 

In AllUris I could already find some interesting hosts. One of them being https://www.shodan.io/host/182.92.132.15. Has a lot of stuff running on it. Could be their server for the app and robots, maybe? On port 8088, I get a JSON response, google translate says it's telling me "Identity verification failed, please login again." Makes it look even more, like this is the home server for the whole operation.

You got some success with MQTT? How network savvy are you? Think you can spoof that host and get in there with mosquito?

I don't like that the whole thing seems to be very Chinese. Wasn't Trifo supposed to be a Californian start-up thing?

**Comment by FrankRizzo890** [comment]
I wrote some code for a cell formation system.  (Converts a bag of chemicals and metal into a battery).  It used MQTT on this little embedded board the size of a stick of gum to communicate with the management server.  I was doing this development work before the management server was built, so I needed something to talk to.  And I used Mosquitto to do that.  It allowed me to monitor what my board was sending, and allowed me to send packet from my Linux box to the board, and see that they were received, and processed correctly.  

The ports TYPICALLY used by MQTT servers are 1883 and 8883, and sometimes 443.  So be on the lookout for those.  If you find them associated with a URL, that's likely the URL for the MQTT server.

**Comment by dscham** [comment]
Well 5687 and 7777 are reporting as ZeroMQ ZMTP 2.0 with nmap. And else, there is only a Dropbear 0.52 SSH server. 

Ooooh, and would look at that. Definitely the right approach, guessing strings and searching for them in the AppStrings page. 🤩 

Found in: com.iot.cloud.sdk.api.fd
```java
public class fd	// class@00119e from classes.dex
{
    private static MqttClient cF;
    private static MqttConnectOptions cG;

    public void fd(){
       super();
    }
    private static void E(){
       String str = "phone/s/10015/1009087183";
       try{
          fd.cF.subscribe(str, 1);
          fd.cF.setCallback(new fe());
       }catch(org.eclipse.paho.client.mqttv3.MqttException e0){
          e0.printStackTrace();
       }
       return;
    }
    private static void F(){
       try{
          MqttClient mqttClient = new MqttClient("ssl://182.92.132.15:10882", MqttClient.generateClientId(), null);
          fd.cF = mqttClient;
          mqttClient.connect(fd.cG);
       }catch(org.eclipse.paho.client.mqttv3.MqttException e0){
          e0.printStackTrace();
       }
       fd.cF.setTimeToWait(0x4e20);
       return;
    }
    private static void G(){
       MqttConnectOptions mqttConnectO = new MqttConnectOptions();
       fd.cG = mqttConnectO;
       mqttConnectO.setCleanSession(true);
       fd.cG.setUserName("iotclientfather");
       fd.cG.setPassword("iot_qinglianyun.com_2015".toCharArray());
       fd.cG.setConnectionTimeout(10);
       fd.cG.setKeepAliveInterval(5);
       fd mqttConnectO1 = new fd();
       InputStream resourceAsSt = fd.class.getResourceAsStream("/cacert.pem");
       try{
          fd.cG.setSocketFactory(SSLUtils.getSocketFactory(resourceAsSt));
       }catch(java.lang.Exception e0){
          e0.printStackTrace();
       }
       return;
    }
    public static void main(String[] p0){
       fd.G();
       fd.F();
       fd.E();
    }
}
```

**Comment by dscham** [comment]
>           MqttClient mqttClient = new MqttClient("ssl://182.92.132.15:10882", MqttClient.generateClientId(), null);
Look who we found again. So that server IS the home server of the whole thing 😏

**Comment by FrankRizzo890** [comment]
Notes from the code above:

          fd.cF.setCallback(new fe());
Looks like it's attaching a new instance of class fe (which probably is just a receiver class) as a callback function.  AKA when a messages arrives, the functions in fe will process it.  (In case you want to know how it handles the messages).

It looks like you have the URL, and this:
       fd.cG.setUserName("iotclientfather");
       fd.cG.setPassword("iot_qinglianyun.com_2015".toCharArray());

It LOOKS like you have everything you need to connect.  

And, in case you haven't found it already:  https://eclipse.dev/paho/files/javadoc/org/eclipse/paho/client/mqttv3/MqttClient.html

**Comment by dscham** [comment]
Yeah, I just downloaded MQTTX, getting a connection refused though. May be the cacert.pem I'm missing. But I couldn't find that yet. Has to be there though, right?

**Comment by dscham** [comment]
I'm going to bed soon. But tomorrow, I'll maybe set up my MikroTik router/ap to route the IP to a VM on my machine and see what it gets.

**Comment by FrankRizzo890** [comment]
> Yeah, I just downloaded MQTTX, getting a connection refused though. May be the cacert.pem I'm missing. But I couldn't find that yet. Has to be there though, right?

fd.class.getResourceAsStream("/cacert.pem");

Yes!  That would certainly suggest that it's a file, and it's called "cacert.pem".

**Comment by dscham** [comment]
Maybe it helps. I got the APK from ApkPure. I don't know the app ecosystem enough, but could it be that there is some files that are put alongside the app during install? If so, then I'd have to get the installed app from my partners phone somehow.

**Comment by FrankRizzo890** [comment]
> Maybe it helps. I got the APK from ApkPure. I don't know the app ecosystem enough, but could it be that there is some files that are put alongside the app during install? If so, then I'd have to get the installed app from my partners phone somehow.

No, not usually!  The APK is a zip file, and is designed to contain everything that the app needs.

**Comment by dscham** [comment]
I invited you to a private repo i created and will push the Trifo Home files there. Also created some Issues to keep track of next steps and or findings.

**Comment by FrankRizzo890** [comment]
Can you slap the APK in there as well?

**Comment by dscham** [comment]
Sure. Gimme a couple minutes

**Comment by FrankRizzo890** [comment]
> Well 5687 and 7777 are reporting as ZeroMQ ZMTP 2.0 with nmap. And else, there is only a Dropbear 0.52 SSH server.
> 
> Ooooh, and would look at that. Definitely the right approach, guessing strings and searching for them in the AppStrings page. 🤩
> 
> Found in: com.iot.cloud.sdk.api.fd
> 
> ```java
> public class fd	// class@00119e from classes.dex
> {
>     private static MqttClient cF;
>     private static MqttConnectOptions cG;
> 
>     public void fd(){
>        super();
>     }
>     private static void E(){
>        String str = "phone/s/10015/1009087183";
>        try{
>           fd.cF.subscribe(str, 1);
>           fd.cF.setCallback(new fe());
>        }catch(org.eclipse.paho.client.mqttv3.MqttException e0){
>           e0.printStackTrace();
>        }
>        return;
>     }
>     private static void F(){
>        try{
>           MqttClient mqttClient = new MqttClient("ssl://182.92.132.15:10882", MqttClient.generateClientId(), null);
>           fd.cF = mqttClient;
>           mqttClient.connect(fd.cG);
>        }catch(org.eclipse.paho.client.mqttv3.MqttException e0){
>           e0.printStackTrace();
>        }
>        fd.cF.setTimeToWait(0x4e20);
>        return;
>     }
>     private static void G(){
>        MqttConnectOptions mqttConnectO = new MqttConnectOptions();
>        fd.cG = mqttConnectO;
>        mqttConnectO.setCleanSession(true);
>        fd.cG.setUserName("iotclientfather");
>        fd.cG.setPassword("iot_qinglianyun.com_2015".toCharArray());
>        fd.cG.setConnectionTimeout(10);
>        fd.cG.setKeepAliveInterval(5);
>        fd mqttConnectO1 = new fd();
>        InputStream resourceAsSt = fd.class.getResourceAsStream("/cacert.pem");
>        try{
>           fd.cG.setSocketFactory(SSLUtils.getSocketFactory(resourceAsSt));
>        }catch(java.lang.Exception e0){
>           e0.printStackTrace();
>        }
>        return;
>     }
>     public static void main(String[] p0){
>        fd.G();
>        fd.F();
>        fd.E();
>     }
> }
> ```

This "fd" code appears to be TEST code.  They install the callback handler (class fe), which has 3 methods, the delivery method does nothing.  (Get called when you SENT it message and the server gets it), there's a disconnect handler which prints a message, and a message received handler which just prints the message it received.  So, that means that the server URL, and user/password might just be for that test server.

**Comment by dscham** [comment]
Yeah, I figured that would be too easy. That server still seems like a really good thing to play with though. Anyway, since it's talking to home and the whole thing is IoT. There has to be some servers. If it's not that one, we'll find another.

I can't push the APK. It's 109 MiB and Github only allows 100 with LFS... Anyway, the extracted-resources, coming up any minute, contains the just plainly extracted APK.

**Comment by FrankRizzo890** [comment]
It's fine.  I wanted it to help find the cert, but since you have it now, there's no need.

**Comment by FrankRizzo890** [comment]
Do you know the entry point for the APK?  (Which class/method gets launched when you click on it)

**Comment by dscham** [comment]
What was the reddit post again? I created a Discord server, maybe we can get some helpers.

> Do you know the entry point for the APK? (Which class/method gets launched when you click on it)
I can find it, I think.

**Comment by dscham** [comment]
> Do you know the entry point for the APK? (Which class/method gets launched when you click on it)

It should be com.trifo.trifohome.App

**Comment by FrankRizzo890** [comment]
I grabbed Trifo Home_2.6.3APKPure, and using GDA you can click the "Entry Point" button (rolls eyes that I didn't remember that), and it takes you to com.trifo.trifohome.view.SplashActivity.OnCreate().  Which makes sense I guess.  Also, have you looked in the "view" directory?  TONS of code that appears to be either NOT obfuscated, or at least VERY LIGHTLY obfuscated.

**Comment by dscham** [comment]
I should be asleep, but that kept me up 😅 

Yeah, I saw that. It's not soo heavily obfuscated. Still, there's some weird stuff going on. Functions that can't be found where they should. And enough obfuscation in the interesting parts IMO.

**Comment by VictorDrijkoningen** [comment]
Wow guys, you've been busy!!

I don't think i fully understand everything, but you've concluded that the mqtt server is for testing right? 
There is a way to turn on device logging on a Trifo Max with the app. Might this be connected? That the test server is for debugging purposes?

>  I created a Discord server, maybe we can get some helpers.

**Comment by dscham** [comment]
I added you to the Repo I created yesterday. Probably smarter to have decompiled code and other assets we get in a private repo at first. And here's the discord https://discord.gg/ZxG7J5cM 😄 

I'm not soo sure yet, that that's only a test server. But we'll see. I'm at my job until the evening. And won't make it as late as yesterday hopefully. But it's a really tempting rabbit hole.
